connectors/slack/template/index.js

```
@@ -1,24 +1,295 @@
+    if (bearerString !== 'Bearer' || !Object.keys(storage.tokenToTeam).includes(token)) {
+      cb(null, {
+        status: 401,
+        body: { message: 'Invalid auth token' },
+      });
+      return;
+    }
+
+    let team, handler;
+    if (!isDeregister) {
+      if (!ctx.body || !ctx.body.app || !ctx.body.team_id || !ctx.body.handler) {
+        cb(null, {
+          status: 401,
+          body: { message: 'Missing required app, team_id, and handler body parameters' },
+        });
+        return;
+      } else {
+        team = `${ctx.body.app}-${ctx.body.team_id}`;
+        handler = ctx.body.handler;
+      }
+    } else {
+      if (!ctx.query || !ctx.query.app || !ctx.query.team_id || !ctx.query.handler) {
+        cb(null, {
+          status: 401,
+          body: { message: 'Missing required app, team_id, and handler query parameters' },
+        });
+        return;
+      } else {
+        team = `${ctx.query.app}-${ctx.query.team_id}`;
+        handler = decodeURIComponent(ctx.query.handler);
+      }
+    }
+
+    const tokenUsedUp = 'TokenUsedUp';
+    const registrationProblem = 'RegistrationProblem';
+
+    storage
+      .put(() => {
+        // Doing the token use check in here in case there is a storage conflict
+        // and when the data is refreshed, it turns out the token was used up
+
+        const [storedTeam, remainingUses] = storage.tokenToTeam[token];
+        if ((!isDeregister && storedTeam !== team) || remainingUses < 1) {
+          Sdk.debug('Token is used up');
+          let e = new Error();
+          e.name = tokenUsedUp;
+          throw e;
+        }
+
+        let handlerRegistered = storage.teamToHandler[team] && storage.teamToHandler[team].includes(handler);
+        if ((!isDeregister && handlerRegistered) || (isDeregister && !handlerRegistered)) {
+          let e = new Error();
+          e.name = registrationProblem;
+          throw e;
```

I wonder if in both cases we should actually silently succeed. In the registration case - to support reentrancy of install.js. In the deregistration case - to support eventual consistency of uninstall.js.

---

Fixed, removed